lms/djangoapps/certificates/api.py

```diff
@@ -425,51 +396,66 @@ def example_certificates_status(course_key):
     return ExampleCertificateSet.latest_status(course_key)
 
 
+def _safe_course_key(course_key):
+    if not isinstance(course_key, CourseKey):
+        try:
+            return CourseKey.from_string(course_key)
+        except InvalidKeyError:
+            log.warning('Unable to parse course_key "%s"', course_key, exc_info=True)
+            return
+    return course_key
+
+
+def _course_from_key(course_key):
+    try:
+        return CourseOverview.get_from_id(_safe_course_key(course_key))
+    except:  # pylint: disable=bare-except
+        log.warning(
+            'Unable to load CourseOverview object for course_key "%s"',
+            unicode(course_key),
+            exc_info=True
+        )
+
+
+def has_html_certificates_enabled(course):
+    if not settings.FEATURES.get('CERTIFICATES_HTML_VIEW', False):
+        return False
+    return course.cert_html_view_enabled
+
+
 def get_certificate_url(user_id=None, course_id=None, uuid=None):
-    """
-    :return certificate url for web or pdf certs. In case of web certs returns either old
-    or new cert url based on given parameters. For web certs if `uuid` is it would return
-    new uuid based cert url url otherwise old url.
-    """
-    url = ""
-    if has_html_certificates_enabled(course_id):
+    course = _course_from_key(course_id)
+    if not course:
+        return ''
+
+    if has_html_certificates_enabled(course):
         if uuid:
-            url = reverse(
+            return reverse(
                 'certificates:render_cert_by_uuid',
                 kwargs=dict(certificate_uuid=uuid)
             )
         elif user_id and course_id:
-            url = reverse(
+            return reverse(
                 'certificates:html_view',
                 kwargs={
                     "user_id": str(user_id),
                     "course_id": unicode(course_id),
                 }
             )
     else:
-        if isinstance(course_id, basestring):
-            try:
-                course_id = CourseKey.from_string(course_id)
-            except InvalidKeyError:
-                log.warning(
-                    ('Unable to parse course_id "%s"', course_id),
-                    exc_info=True
-                )
-                return url
         try:
             user_certificate = GeneratedCertificate.eligible_certificates.get(
                 user=user_id,
-                course_id=course_id
+                course_id=_safe_course_key(course_id)
             )
-            url = user_certificate.download_url
+            return user_certificate.download_url
         except GeneratedCertificate.DoesNotExist:
             log.critical(
                 'Unable to lookup certificate\n'
                 'user id: %d\n'
                 'course: %s', user_id, unicode(course_id)
             )
-
-    return url
+    return ''
```

Comment:
I generally prefer having single `return` statements in a method/function - it helps with debugging and readability.  So I'm curious why this change was made - but I don't consider it a must/blocker.

Reply:
I dunno, I find early returns to be slightly easier to reason about.  I'll try it the other way to humor you, though.